src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.ObsoletedInOSPlatformTests.cs

```diff
@@ -0,0 +1,457 @@
+                    AttributeTargets.Field |
+                    AttributeTargets.Interface |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple = true, Inherited = false)]
+    public sealed class UnsupportedOSPlatformAttribute : OSPlatformAttribute
+    {
+        public UnsupportedOSPlatformAttribute(string platformName) : base(platformName)
+        {
+        }
+        public UnsupportedOSPlatformAttribute(string platformName, string message) : base(platformName)
+        {
+            Message = message;
+        }
+        public string Message { get; }
+    }
+
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Enum |
+                    AttributeTargets.Event |
+                    AttributeTargets.Field |
+                    AttributeTargets.Interface |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple = true, Inherited = false)]
+    public sealed class ObsoletedInOSPlatformAttribute : OSPlatformAttribute
+    {
+        public ObsoletedInOSPlatformAttribute(string platformName) : base(platformName)
+        {
+        }
+        public ObsoletedInOSPlatformAttribute(string platformName, string message) : base(platformName)
+        {
+            Message = message;
+        }
+        public string Message { get; }
+        public string Url { get; set; }
+    }
+}";
+
+        private readonly string MockObsoletedAttributeVB = @"
+Namespace Mock
+    <AttributeUsage(AttributeTargets.Assembly Or
+                    AttributeTargets.[Class] Or AttributeTargets.Constructor Or
+                    AttributeTargets.[Enum] Or AttributeTargets.[Event] Or
+                    AttributeTargets.Field Or AttributeTargets.[Interface] Or
+                    AttributeTargets.Method Or AttributeTargets.[Module] Or
+                    AttributeTargets.[Property] Or
+                    AttributeTargets.Struct,
+                    AllowMultiple:=True, Inherited:=False)>
+    Public NotInheritable Class ObsoletedInOSPlatformAttribute
+        Inherits Attribute
+
+        Public Sub New(ByVal platformName As String)
+            PlatformName = platformName
+        End Sub
+
+        Public Sub New(ByVal platformName As String, ByVal message As String)
+            PlatformName = platformName
+            Message = message
+        End Sub
+
+        Public ReadOnly Property PlatformName As String
+        Public ReadOnly Property Message As String
+        Public Property Url As String
+    End Class
+End Namespace";
+    }
+}
```

Comment:
I recommend adding some tests that illustrate the common scenario for ios platforms where:

1. A minimum version is specified for `[SupportedOSPlatform]`
2. A later version is specified for `[ObsoletedInOSPlatform]`
3. A later version is specified for `[UnsupportedOSPlatform]`

Another test case that could be valuable is for checking the logic around the ObsoletedInOSPlatform version being outside the bounds of the supported/unsupported range (in both directions).

Reply:
> Another test case that could be valuable is for checking the logic around the ObsoletedInOSPlatform version being outside the bounds of the supported/unsupported range (in both directions).

I believe that should apply for callsite only, that test exist: https://github.com/dotnet/roslyn-analyzers/blob/18f5a444bac98eec105e90d0376e8cacb0297756/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.ObsoletedInOSPlatformTests.cs#L103-L104

But the other scenario is missing, I will add a test for that

Reply:
Ah, OK -- so that test covers if the call site's applicable range does not include the obsolete version; right?

Is there value in testing something like this:

`
[SupportedOSPlatform("ios10.0")]
[UnsupportedOSPlatform("ios12.0")]
[ObsoletedInOSPlatform("ios14.0")]
public void ObsoletedAfterUnsupported() { }
`

With call sites reachable on 9.0, 11.0, 13.0, and 15.0?

Reply:
While adding the missing scenario found a that obsoleted was to handled properly for parent child members attributes relation, thanks! Also added tests for `ObsoletedAfterUnsupported` API.

> With call sites reachable on 9.0, 11.0, 13.0, and 15.0?

Different support version for callsite has no difference, because  reachable on 9.0, 11.0, 13.0, and 15.0 means reachable on 9.0+, 11.0+, 13.0+, and 15.0+, same for obsoleted, therefore from which ever version the API is supported we should warn for any obsoleted version because they would overlapped eventually, only unsupported version could make difference for obsoleted, and I have added tests for that